packages/platform-android/src/commands/runAndroid/runOnEmulator.ts

```diff
@@ -26,29 +26,62 @@ function toPascalCase(value: string) {
 
 type AndroidProject = NonNullable<Config['project']['android']>;
 
-async function runOnAllDevices(
+function logStartEmulatorError(error: Error) {
+  if (error.message) {
+    logger.error(
+      `Failed to launch emulator. Reason: ${chalk.dim(error.message)}`,
+    );
+  } else {
+    logger.error('Failed to launch emulator.');
+  }
+
+  logger.warn(
+    'Please launch an emulator manually or connect a device. Otherwise the app may fail to launch.',
+  );
+}
+
+async function startEmulator(adbPath: string, deviceName?: string) {
+  if (deviceName) {
+    logger.info('Launching emulator...');
+
+    try {
+      const deviceListAfterEmulatorLaunched = await launchEmulator(
+        deviceName,
+        adbPath,
+      );
+
+      return deviceListAfterEmulatorLaunched;
+    } catch (error) {
+      return logStartEmulatorError(error);
+    }
+  }
+
+  const devices = adb.getDevices(adbPath);
+  if (devices.length > 0) {
+    return devices;
+  }
+
+  logger.info('Launching emulator...');
+
+  try {
+    const deviceListAfterAnyEmulatorLaunched = await launchAnyEmulator(adbPath);
+
+    return deviceListAfterAnyEmulatorLaunched;
+  } catch (error) {
+    console.log({error});
+    return logStartEmulatorError(error);
+  }
+}
+
+async function runOnEmulator(
   args: Flags,
   cmd: string,
   packageName: string,
   adbPath: string,
   androidProject: AndroidProject,
+  deviceName?: string,
 ) {
-  let devices = adb.getDevices(adbPath);
-  if (devices.length === 0) {
-    logger.info('Launching emulator...');
-    const result = await tryLaunchEmulator(adbPath);
-    if (result.success) {
-      logger.info('Successfully launched emulator.');
-      devices = adb.getDevices(adbPath);
-    } else {
-      logger.error(
-        `Failed to launch emulator. Reason: ${chalk.dim(result.error || '')}.`,
-      );
-      logger.warn(
-        'Please launch an emulator manually or connect a device. Otherwise app may fail to launch.',
-      );
-    }
-  }
+  const devices = await startEmulator(adbPath, deviceName);
```

Comment:
I would prefer to stick to previous:
```js
await startEmulator(adbPath, deviceName);
const devices = adb.getDevices(adbPath);
```
instead of passing around a list of devices, which makes the code a bit harder to follow.

Alternatively, I would also accept:
```js
const launchedDevice = await startEmulator(adbPath, deviceName);
```
but I don't think it's useful, as we have to run `adb.getDevices()` anyway.

Reply:
When we address that, the code will become much easier in other places, due to lack of `return`